src/EncodingMsg.chpl

```diff
@@ -9,38 +9,61 @@ module EncodingMsg {
     use CommAggregation;
     use ServerErrors;
     use ServerErrorStrings;
+    use Codecs;
+    use CTypes;
+
+    use AryUtil;
 
     use SegmentedString;
 
     private config const logLevel = ServerConfig.logLevel;
     const emLogger = new Logger(logLevel);
 
     proc encodeDecodeMsg(cmd: string, msgArgs: borrowed MessageArgs, st: borrowed SymTab): MsgTuple throws {
-        var repMsg: string;
-        var encoding = msgArgs.getValueOf("encoding");
-
-        var stringsObj = getSegString(msgArgs.getValueOf("obj"), st);
-
-        select encoding.toLower() {
-            when "idna" {
-                var (offsets, vals) = stringsObj.idnaEncodeDecode(cmd); // use the cmd to trigger correct action
-                var encodedStrings = getSegString(offsets, vals, st);
-                repMsg = "created " + st.attrib(encodedStrings.name) + "+created bytes.size %t".format(encodedStrings.nBytes);
-                
-            }
-            otherwise {
-                var errorMsg = "%s %s not currently supported".format(cmd, encoding);      
-                emLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);  
-                return new MsgTuple(errorMsg, MsgType.ERROR);    
-            }
-        }
-
-        emLogger.debug(getModuleName(), getRoutineName(), getLineNumber(), repMsg);
-        return new MsgTuple(repMsg, MsgType.NORMAL);
+      var repMsg: string;
+
+      var toEncoding = msgArgs.getValueOf("toEncoding").toUpper();
+      var fromEncoding = msgArgs.getValueOf("fromEncoding").toUpper();
+      
+      var stringsObj = getSegString(msgArgs.getValueOf("obj"), st);
+
+      var (offsets, vals) = encodeDecode(stringsObj, toEncoding, fromEncoding);
+      var encodedStrings = getSegString(offsets, vals, st);
+      repMsg = "created " + st.attrib(encodedStrings.name) + "+created bytes.size %t".format(encodedStrings.nBytes);
+
+      emLogger.debug(getModuleName(), getRoutineName(), getLineNumber(), repMsg);
+      return new MsgTuple(repMsg, MsgType.NORMAL);
     }
 
+    proc encodeDecode(stringsObj, toEncoding: string, fromEncoding: string) throws {
+      ref origVals = stringsObj.values.a;
+      ref offs = stringsObj.offsets.a;
+      // TODO: is this inefficient?
+      const lengths = stringsObj.getLengths();
+
+      // TODO: We only need one of these, they have the same info
+      var encodeLengths: [stringsObj.offsets.a.domain] int;
+      var encodeOffsets: [stringsObj.offsets.a.domain] int;
+      
+      // First get lengths
+      forall (i, off, len) in zip(0..#stringsObj.size, offs, lengths) {
+        var slice = new lowLevelLocalizingSlice(origVals, off..#len);
+        encodeLengths[i] = getBufLength(slice.ptr: c_ptr(uint(8)), len, toEncoding, fromEncoding);
+      }
```

Comment:
if we try to optimize more, we might be able to take advantage of the `computeOnSegments` for this loop. I'm not sure if it will be more or less efficient but might be worth looking into

Reply:
Ya, I have some optimizations that are in the works and getting rid of this `lowLevelLocalizing` slice is the first one on the list. Thanks for pointing out `computeOnSegments`, that is not something I'm familiar with and I'll investigate.